---
title: "RNA-seq transcriptomics data"
author: "Yuying Shi"
date: "2024-07-16"
output: 
  pdf_document:
    latex_engine: xelatex
    
mainfont: "SimSun"
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Introduction of dataset

The 'data_GC' dataset is used for RNA-seq transcriptomics data, sourced from the publicly accessible database GEO PRJNA152559, GSE35809. It includes genome-wide mRNA expression profiles of 70 primary gastric tumors from an Australian patient cohort, categorized into metabolic (15 samples), invasive (26 samples), and proliferative subtypes (29 samples).We used proliferative gastric cancer samples as the experimental group and the other two groups of samples as the control group to demonstrate the process of genetic difference analysis.

```{r data_GC}
load("Rmarkdown_gastric_cancer.RData")
head(data_GC,3)
```

## differential genes expression analysis

Genes with P value less than 0.05 and \|logFC\| value greater than or equal to 1 were taken as differential expression genes.

```{r data_GC -differential genes expression analysis}

library(tidyr)
library(limma)

new_GC <- t(data_GC[,-1])
group_GC <- data_GC[,1]

new_GC <- new_GC %>% data.frame(stringsAsFactors = FALSE) %>%
  data.matrix()
  
design <- model.matrix(~0 + factor(group_GC))
colnames(design) <- c("control", "treat")

rownames(design) <- colnames(new_GC)
contrast.matrix <- makeContrasts(treat - control, levels = design)

fit <- lmFit(new_GC, design)
fit_2 <- eBayes(contrasts.fit(fit, contrast.matrix))
temoutput <- limma::topTable(fit_2, n = Inf, adjust = "fdr")
temoutput <- na.omit(temoutput)

temoutput <- dplyr::mutate(temoutput, g = ifelse(temoutput$adj.P.Val < 0.05 & abs(temoutput$logFC) >= 1,ifelse(temoutput$logFC >1, 'Up', 'Down'),'Stable'))

temoutput
table(temoutput$g)
```

```{r volcano plot}

library(ggplot2)

professional_colors <- c("blue", "black", "red")

ggplot(temoutput, aes(x=logFC, y=-log10(adj.P.Val))) +
  geom_hline(yintercept = -log10(0.05), linetype = 'dashed', size = 0.5, color = "grey60") +
  geom_vline(xintercept = c(-1,1), linetype = 'dashed', size = 0.5, color = "grey60") +
  geom_point(aes(color = g), size = 2.5, alpha = 0.7) +
  scale_color_manual(values = professional_colors) +
  labs(title = "Volcano Plot", x = "Log Fold Change (logFC)", y = "-Log10 P-value", colour = "Group") +
  # theme_classic(base_family = "Arial") +
  theme(
    text = element_text(size = 16),
    plot.title = element_text(size = 28, hjust = 0.5, margin = margin(b = 80, unit = "pt")),
    axis.title = element_text(size = 22),
    axis.text = element_text(size = 18),
    legend.position = "right",
    legend.title = element_text(size = 18),
    legend.text = element_text(size = 16),
    legend.key.size = unit(3, "lines"),
    plot.margin = margin(1, 1, 1, 1, "cm"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  )

```

```         
```

```{r KEGG pathway analysis}
genes <- rownames(temoutput[temoutput$g %in% c("Up","Down"),]) %>% data.frame()
colnames(genes) <- "gene"

library(clusterProfiler)  
library(org.Hs.eg.db)
diffgene<-bitr(genes$gene,fromType = "SYMBOL",toType = "ENTREZID",OrgDb = "org.Hs.eg.db")
genes_entrez <- diffgene$ENTREZID
enrich_result <- enrichKEGG(gene = genes_entrez,
                            organism = "hsa",
                            keyType = "kegg", 
                            pAdjustMethod = "holm",
                            pvalueCutoff = 0.05,
                            qvalueCutoff = 0.05)
data.frame(enrich_result)
```

```{r KEGG pathway plot}
ek.rt <- data.frame(enrich_result@result)

ek.rt <- separate(data=ek.rt, col=GeneRatio, into = c("GR1", "GR2"), sep = "/") #劈分GeneRatio为2列（GR1、GR2）
ek.rt <- separate(data=ek.rt, col=BgRatio, into = c("BR1", "BR2"), sep = "/") #劈分BgRatio为2列（BR1、BR2）
ek.rt <- mutate(ek.rt, enrichment_factor = (as.numeric(GR1)/as.numeric(GR2))/(as.numeric(BR1)/as.numeric(BR2))) #计算Enrichment Factor

ek.rt10 <- ek.rt %>% dplyr::filter(dplyr::row_number() >= 1,dplyr::row_number() <= as.numeric(20))

p1 <-  ggplot(ek.rt10, aes(enrichment_factor, Description)) + 
  geom_point(aes(size=Count, color=p.adjust)) +
  scale_color_gradient(low="red", high="blue") + 
  labs(color="P Adjust", size="Count", x="Enrichment Factor", y="KEGG Term", title="KEGG Enrichment Analysis") + 
  theme_minimal() +
  theme(
    text = element_text(size=10), 
    axis.title = element_text(size=10, face="bold"), 
    axis.text = element_text(size=10), 
    legend.title = element_text(size=10), 
    legend.text = element_text(size=10),
    axis.text.x = element_text(size = 10),  
    axis.text.y = element_text(size = 10),
    plot.title = element_text(size=14, face="bold", hjust=0.5))
p1
 
p2 <-  ggplot(ek.rt10, aes(x = enrichment_factor, y = reorder(Description, enrichment_factor))) + 
   geom_bar(stat = "identity", aes(fill = p.adjust), width = 0.7) +
   scale_fill_gradient(low = "red", high = "blue", name = "P Adjust") + 
   labs(size = "Count", x = "Enrichment Factor", y = "KEGG Term", title = "KEGG Enrichment Analysis") + 
   theme_minimal() +
   theme(
     text = element_text(size = 10), 
     axis.title = element_text(size = 10, face = "bold"), 
     axis.text = element_text(size = 10), 
     legend.title = element_text(size = 10), 
     legend.text = element_text(size = 10),
     axis.text.x = element_text(size = 10),  
     axis.text.y = element_text(size = 10),
     plot.title = element_text(size = 14, face = "bold", hjust = 0.5)
   )
p2
 
```
